plugins/net_plugin/net_plugin.cpp

```diff
@@ -1097,32 +1100,38 @@ namespace eosio {
       if( ec ) {
          fc_elog( logger, "connection failed (set_option) ${peer}: ${e1}", ("peer", peer_name())( "e1", ec.message() ) );
          close();
-         return false;
+         callback(false);
       } else {
 
-         auto start_read = [this](){
-            fc_dlog( logger, "connected to ${peer}", ("peer", peer_name()) );
-            socket_open = true;
-            start_read_message();
-         };
          if (my_impl->ssl_enabled) {
             socket.ssl_socket()->async_handshake(server ? ssl::stream_base::server : ssl::stream_base::client,
-                                       boost::asio::bind_executor(strand, [start_read, c = shared_from_this(), socket=socket](const auto& ec){
+                                       boost::asio::bind_executor(strand, 
+                                       [c = shared_from_this(), socket=socket, callback](const auto& ec){
                                           //we use socket just to retain connection shared_ptr just in case it will be deleted
                                           //when we will need it inside start_read_message
                                           std::ignore = socket;
                                           if (ec) {
+                                             c->set_participating(false);
                                              fc_elog( logger, "ssl handshake error: ${e}", ("e", ec.message()) );
                                              c->close();
                                              return;
                                           }
-                                          start_read();
+
+                                          c->set_socket_open();
+                                          c->set_participating(true);
+                                          fc_dlog( logger, "connected to ${peer}", ("peer", c->peer_name()) );
+                                          fc_dlog( logger, "participant added: ${o}", ("o", c->participant_name()));
+                                          
+                                          c->start_read_message();
+                                          callback(true);
                                        }));
          }
          else {
-            start_read();
+            fc_dlog( logger, "connected to ${peer}", ("peer", peer_name()) );
```

Comment:
is there a reason that we couldn't still have this in a lambda, that both this path and the ssl path use? they still do exactly the same actions, except the ssl path has an extra log of "participant added...".

Reply:
yes, we can, but we need to handle somehow `set_participating` logic that is unnecessary in case of non-ssl. I'll think of it.

Reply:
can you call c->set_participating(true) before calling the lambda, or is that problematic? This isn't absolutely necessary, but I like that it makes it clear that the same logic is happening for both (just later on for SSL)